### PR TITLE
Overall improvements in modularity 

### PR DESCRIPTION
**Pull request by aanm**
I had these commits on my side for a couple weeks now, they are not that coupled but it wouldn't make sense to open 4 PRs for such small changes. The overall goal is to make the code more modular and depend less on some imports. Read on per commit basis.

### CODE REVIEW

**Review by joamaki — APPROVED** [review]

File: pkg/endpoint/endpoint.go [hunk 1]
```diff
@@ -509,8 +509,8 @@ func (e *Endpoint) waitForProxyCompletions(proxyWaitGroup *completion.WaitGroup)
 	return nil
 }
 
-// NewEndpointWithState creates a new endpoint useful for testing purposes
-func NewEndpointWithState(owner regeneration.Owner, policyGetter policyRepoGetter, namedPortsGetter namedPortsGetter, proxy EndpointProxy, allocator cache.IdentityAllocator, ID uint16, state State) *Endpoint {
+// NewTestEndpointWithState creates a new endpoint useful for testing purposes
+func NewTestEndpointWithState(owner regeneration.Owner, policyGetter policyRepoGetter, namedPortsGetter namedPortsGetter, proxy EndpointProxy, allocator cache.IdentityAllocator, ID uint16, state State) *Endpoint {
```
Comment (pippolo84) [inline pkg/endpoint/endpoint.go]: To make it even more explicit that this should be used for testing purposes only, we can add a `testing.TB` as a first parameter. Doing so, we'll be able to call this from:

- a gocheck test with `NewTestEndpointWithState(c, ...)`
- a Go std test with: `NewTestEndpointWithState(t, ...)`
- a Go benchmark with `NewTestEndpointWithState(b, ...)`

File: api/v1/openapi.yaml [hunk 2]
```diff
@@ -1545,6 +1545,12 @@ definitions:
         type: boolean
       datapath-configuration:
         "$ref": "#/definitions/EndpointDatapathConfiguration"
+      metadata:
+        description: >-
+          Metadata used in to store information about the endpoint at creation.
```
Comment (pippolo84) [inline api/v1/openapi.yaml]: typo: `in` should be removed

**Review by pippolo84 — APPROVED** [review]
Thanks! :100: 
Left a couple of non-blocking nits inline.

File: pkg/endpoint/endpoint.go (hunk 1 above)
Comment (aanm) [reply pkg/endpoint/endpoint.go]: good idea!

**Review by jibi — APPROVED** [review]

File: pkg/endpoint/endpoint.go [hunk 3]
```diff
@@ -385,6 +403,9 @@ type Endpoint struct {
 	// mutable! must hold the endpoint lock to read
 	ciliumEndpointUID k8sTypes.UID
 
+	// metadata is used to store some internal metadata about this Endpoint.
+	metadata map[string]interface{}
```
Comment (squeed) [inline pkg/endpoint/endpoint.go]: naming nit: this isn't metadata, it's data. In other words, it directly influences the endpoint's configuration, same as anything else. Plus, metadata already implicitly means name + namespace in the k8s world.

What about calling this TestTweaks? Or something like that?

File: pkg/endpoint/endpoint.go [hunk 4]
```diff
@@ -1321,6 +1321,38 @@ func (e *Endpoint) GetPod() *slim_corev1.Pod {
 	return e.pod.Load()
 }
 
+// CEPOwnerInterface contains the interface of an endpoint owner.
+type CEPOwnerInterface interface {
```
Comment (squeed) [inline pkg/endpoint/endpoint.go:1328]: What about using the k8s [Object interface](https://pkg.go.dev/k8s.io/apimachinery@v0.29.1/pkg/apis/meta/v1#Object) instead?

File: pkg/endpoint/endpoint.go [hunk 5]
```diff
@@ -1321,6 +1321,38 @@ func (e *Endpoint) GetPod() *slim_corev1.Pod {
...
+	GetUID() k8sTypes.UID
+
+	// GetHostIP returns the owners' host IP.
+	GetHostIP() string
+}
+
+// GetCEPOwner retrieves the pod related to this endpoint
+func (e *Endpoint) GetCEPOwner() CEPOwnerInterface {
+	return e.pod.Load()
+}
```
Comment (aditighag) [inline pkg/endpoint/endpoint.go]: The name is confusing as `GetPod` above also returns `e.pod.Load()`.

**Review by aditighag — CHANGES_REQUESTED** [review]
>introduce 'metadata' for endpoints

I share same concerns as Casey for defining "metadata". Also, the PR is replacing global configs with an embedded per endpoint property. The commit message mentions unit tests as the motivation, but it doesn't mention what the problem(s) are.

File: pkg/endpoint/endpoint.go (hunk 4 above)
Comment (aanm) [reply pkg/endpoint/endpoint.go:1328]: > What about using the k8s [Object interface](https://pkg.go.dev/k8s.io/apimachinery@v0.29.1/pkg/apis/meta/v1#Object) instead?

The issue is that it has way more methods that it's required as well as it doesn't provide the `IsNil()` method.

**Review by nathanjsweet — APPROVED** [review]
Looks good.

File: pkg/endpoint/endpoint.go (hunk 3 above)
Comment (aditighag) [reply pkg/endpoint/endpoint.go]: Or maybe `properties`?

File: pkg/endpoint/endpoint.go [hunk 6]
```diff
@@ -509,10 +531,11 @@ func (e *Endpoint) waitForProxyCompletions(proxyWaitGroup *completion.WaitGroup)
 	return nil
 }
 
-// NewEndpointWithState creates a new endpoint useful for testing purposes
-func NewEndpointWithState(owner regeneration.Owner, policyGetter policyRepoGetter, namedPortsGetter namedPortsGetter, proxy EndpointProxy, allocator cache.IdentityAllocator, ID uint16, state State) *Endpoint {
+// NewTestEndpointWithState creates a new endpoint useful for testing purposes
+func NewTestEndpointWithState(_ testing.TB, owner regeneration.Owner, policyGetter policyRepoGetter, namedPortsGetter namedPortsGetter, proxy EndpointProxy, allocator cache.IdentityAllocator, ID uint16, state State) *Endpoint {
```
Comment (aditighag) [inline pkg/endpoint/endpoint.go:535]: I checked all the references, and it's only been called in unit tests. Move it to testutils? It feels rather awkward to leave this in `pkg/endpoint`.

File: pkg/endpoint/bpf.go [hunk 7]
```diff
@@ -202,7 +202,7 @@ func (p *proxyPolicy) GetListener() string {
 // Only called after a new desired policy has been computed.
 // Must be called with endpoint.mutex Lock()ed.
 func (e *Endpoint) addNewRedirectsFromDesiredPolicy(ingress bool, desiredRedirects map[string]uint16, proxyWaitGroup *completion.WaitGroup) (error, revert.FinalizeFunc, revert.RevertFunc) {
-	if option.Config.DryMode || e.IsProxyDisabled() {
+	if e.isMetadata(MetadataFakeEndpoint) || e.IsProxyDisabled() {
```
Comment (aditighag) [inline pkg/endpoint/bpf.go]: This comment applies to all the below changes - 

I would suggest adding a wrapper that checks `option.Config.DryMode` or whether an endpoint is a fake one. This way, we preserve the existing behavior.

File: pkg/endpoint/bpf.go [hunk 8]
```diff
@@ -543,12 +543,12 @@ func (e *Endpoint) regenerateBPF(regenContext *regenerationContext) (revnum uint
...
 
 	// No need to compile BPF in dry mode.
-	if option.Config.DryMode {
+	if e.isMetadata(MetadataFakeEndpoint) {
 		return e.nextPolicyRevision, false, nil
 	}
 
 	// Skip BPF if the endpoint has no policy map
-	if !e.HasBPFPolicyMap() {
+	if e.isMetadata(MetadataSkipBPFPolicy) {
```
Comment (aditighag) [inline pkg/endpoint/bpf.go]: Same comment as above (L 205), but for `HasBPFPolicyMap`.

File: api/v1/openapi.yaml [hunk 9]
```diff
@@ -1545,6 +1545,12 @@ definitions:
         type: boolean
       datapath-configuration:
         "$ref": "#/definitions/EndpointDatapathConfiguration"
+      metadata:
+        description: >-
+          Metadata used to store information about the endpoint at creation.
```
Comment (aditighag) [inline api/v1/openapi.yaml]: Metadata -> properties?

File: pkg/endpoint/endpoint.go (hunk 3 above)
Comment (aditighag) [reply pkg/endpoint/endpoint.go]: Or maybe `properties`?

File: pkg/endpoint/endpoint.go [hunk 10]
```diff
@@ -1294,6 +1321,38 @@ func (e *Endpoint) GetPod() *slim_corev1.Pod {
 	return e.pod.Load()
 }
 
+// CEPOwnerInterface contains the interface of an endpoint owner.
```
Comment (aditighag) [inline pkg/endpoint/endpoint.go:1327]: How about adding a note: by default it's the pod information associated with an endpoint? There is no other reference to endpoint owner in the code.

**Review by aditighag — CHANGES_REQUESTED** [review]
Synced up offline. My main concern is about changing the existing behavior. Please see the suggestions inline.

File: pkg/endpoint/endpoint.go (hunk 6 above)
Comment (aanm) [reply pkg/endpoint/endpoint.go:535]: Yeah, the problem is that if I move to testutils, it doesn't have access to the internal fields of the endpoint.

File: api/v1/openapi.yaml (hunk 9 above)
Comment (aanm) [reply api/v1/openapi.yaml]: Oh nice name!

File: pkg/endpoint/endpoint.go (hunk 3 above)
Comment (aanm) [reply pkg/endpoint/endpoint.go]: I'm going with properties.

File: pkg/endpoint/bpf.go (hunk 7 above)
Comment (aanm) [reply pkg/endpoint/bpf.go]: The overall idea on the commit is also to stop relying on `option.Config.DryMode` with this change.

File: pkg/endpoint/bpf.go (hunk 8 above)
Comment (aanm) [reply pkg/endpoint/bpf.go]: I've marked `HasBPFPolicyMap` as deprecated.

File: pkg/endpoint/endpoint.go [hunk 11]
```diff
@@ -579,6 +603,9 @@ func CreateIngressEndpoint(owner regeneration.Owner, policyGetter policyRepoGett
 	ep.DatapathConfiguration = NewDatapathConfiguration()
 
 	ep.isIngress = true
+	ep.properties[PropertySkipBPFPolicy] = true
+	ep.properties[PropertyWithouteBPFDatapath] = true
```
Comment (aditighag) [inline pkg/endpoint/endpoint.go:610]: Is this change intentional? Can you add a comment?
How was this handled before this change?

**Review by aditighag — CHANGES_REQUESTED** [review]
Thanks for the revisions. 
Q around how `PropertySkipBPFRegeneration` and `PropertyFakeEndpoint` are set, and under what conditions.

File: pkg/endpoint/endpoint.go (hunk 11 above)
Comment (aanm) [reply pkg/endpoint/endpoint.go:610]: > Is this change intentional? Can you add a comment?

It is intentional, I forgot to add a comment.

This was handled before with the `ep.isIngress = true`.

**Review by aditighag — APPROVED** [review]
